### ISSUE TITLE
Flreconstruct not working in Falaise 3 for custom modules

### ISSUE THREAD

**Issue by noeroy** [SuperNEMO-DBD/Falaise#7]
Hello,
I have followed the instructions on the Falaise 3 Documentations to write a custom module, but when I try to run Flreconstruct with the module proposed in the tutorial, I get this error : 

[noeroy@pc179 MyModule-build]$ /unix/nemo2/users/cpatrick/CadfaelBrew/bin/flreconstruct -i /home/noeroy/BRIO/test2.brio -p /home/noeroy/Modules/MyModule/MyModulePipeline.conf -o output.root

Error in cling::AutoloadingVisitor::InsertIntoAutoloadingState:
   Missing FileEntry for /tmp/bayeux-20170515-27376-1qs7r7s/Bayeux-3.0.0/source/bxbrio/include/brio/detail/brio_record.h
   requested to autoload type TArrayCMod
Error in cling::AutoloadingVisitor::InsertIntoAutoloadingState:
   Missing FileEntry for /tmp/bayeux-20170515-27376-1qs7r7s/Bayeux-3.0.0/source/bxbrio/include/brio/detail/brio_record.h
   requested to autoload type brio_record
[fatal:falaise::exit_code FLReconstruct::do_pipeline(const FLReconstruct::FLReconstructParams&):155] Failed to initialize pipeline : [void dpp::module_manager::_create_module(dpp::module_entry_type&):279: Cannot find module factory with ID 'MyModule' for module named 'pipeline' !]

**Comment by cherylepatrick** [comment]
In case this helps with tracking it down, we've seen this on both Linux and Mac. Noe saw the error when following the instructions on the tutorial website, and I get the same error when running my SensitivityModule custom module, which used to work fine (on both mac and linux) with the old Falaise.

**Comment by drbenmorgan** [comment]
@noeroy , @cherylepatrick Just to confirm, could you post the config scripts being run, and also:

- where the `.so/.dylib` file for the module is located
- where `flreconstruct` is being run from (i.e. the current working directory)

It looks like the module library isn't being located/loaded.

**Comment by cherylepatrick** [comment]
On my mac for the SensitivityModule:
```
eduroam-int-dhcp-97-252-40:testfalaise3 cpatrick$ pwd
/Users/cpatrick/SuperNEMO/testfalaise3
eduroam-int-dhcp-97-252-40:testfalaise3 cpatrick$ flreconstruct -i testflrec.brio -p /Users/cpatrick/Dropbox/SuperNEMO/ucl_svn/SensitivityStudy/SensitivityModule.conf
Error in cling::AutoloadingVisitor::InsertIntoAutoloadingState:
   Missing FileEntry for /tmp/bayeux-20170510-77836-4va8dm/Bayeux-3.0.0/source/bxbrio/include/brio/detail/brio_record.h
   requested to autoload type TArrayCMod
Error in cling::AutoloadingVisitor::InsertIntoAutoloadingState:
   Missing FileEntry for /tmp/bayeux-20170510-77836-4va8dm/Bayeux-3.0.0/source/bxbrio/include/brio/detail/brio_record.h
   requested to autoload type brio_record
[fatal:falaise::exit_code FLReconstruct::do_pipeline(const FLReconstruct::FLReconstructParams &):155] Failed to initialize pipeline : [void dpp::module_manager::_create_module(dpp::module_entry_type &):279: Cannot find module factory with ID 'SensitivityModule' for module named 'processing' !]
```
The `dylib` is in `/Users/cpatrick/Dropbox/SuperNEMO/ucl_svn/SensitivityStudy/SensitivityModule-build`

And lastly here is the `SensitivityModule.conf` file:

```
# - Configuration Metadata
#@description Chain pipeline using a single custom module
#@key_label   "name"
#@meta_label  "type"
# - Custom modules
# The "flreconstruct.plugins" section to tell flreconstruct what
# to load and from where.
[name="flreconstruct.plugins" type=""]
plugins : string[1] = "SensitivityModule"
SensitivityModule.directory : string = "/Users/cpatrick/Dropbox/SuperNEMO/ucl_svn/SensitivityStudy/SensitivityModule-build"
# - Pipeline configuration
# Must define "pipeline" as this is the module flreconstruct will use
# Make it use our custom module by setting the'type' key to the string we
# used as the second argument to the macro
# DPP_MODULE_REGISTRATION_IMPLEMENT in SensitivityModule.cpp
# At present, it takes no configuration, so it suffices to define it
[name="pipeline" type="dpp::chain_module"]
modules : string[1] = "processing"
#modules : string[2] = "processing" "resultdump"
#[name="resultdump" type="dpp::dump_module"]
#title : string = "ResultDump"
[name="processing" type="SensitivityModule"]
```

Thanks once again @drbenmorgan , I know you are spending half your life running around after my stupid error messages these days and I do appreciate it.

**Comment by drbenmorgan** [comment]
@cherylepatrick Thanks! Nothing stupid about error messages - it's incredibly useful to have this feedback to improve the software, so keep it coming :-)

**Comment by noeroy** [comment]
The `.so` is in `/home/noeroy/Modules/MyModule/MyModule-build` and I have got that for flreconstruct : 

```
[noeroy@pc179 MyModule-build]$ pwd
/home/noeroy/Modules/MyModule/MyModule-build
[noeroy@pc179 MyModule-build]$ /unix/nemo2/users/cpatrick/CadfaelBrew/bin/flreconstruct -i /home/noeroy/BRIO/test2.brio -p /home/noeroy/Modules/MyModule/MyModulePipeline.conf 
Error in cling::AutoloadingVisitor::InsertIntoAutoloadingState:
   Missing FileEntry for /tmp/bayeux-20170515-27376-1qs7r7s/Bayeux-3.0.0/source/bxbrio/include/brio/detail/brio_record.h
   requested to autoload type TArrayCMod
Error in cling::AutoloadingVisitor::InsertIntoAutoloadingState:
   Missing FileEntry for /tmp/bayeux-20170515-27376-1qs7r7s/Bayeux-3.0.0/source/bxbrio/include/brio/detail/brio_record.h
   requested to autoload type brio_record
[fatal:falaise::exit_code FLReconstruct::do_pipeline(const FLReconstruct::FLReconstructParams&):155] Failed to initialize pipeline : [void dpp::module_manager::_create_module(dpp::module_entry_type&):279: Cannot find module factory with ID 'MyModule' for module named 'pipeline' !]

```
Here is the `MyModulePipeline.conf` : 

```
# - Configuration Metadata
#@description Chain pipeline using a single custom module
#@key_label   "name"
#@meta_label  "type"
# - Custom modules
# The "flreconstruct.plugins" section to tell flreconstruct what
# to load and from where.
[name="flreconstruct.plugins" type=""]
plugins : string[1] = "MyModule"
MyModule.directory : string = "/home/noeroy/Modules/MyModule/MyModule-build"
# - Pipeline configuration
# Must define "pipeline" as this is the module flreconstruct will use
# Make it use our custom module by setting the'type' key to the string we
# used as the second argument to the macro
# DPP_MODULE_REGISTRATION_IMPLEMENT in MyModule.cpp
# At present, it takes no configuration, so it suffices to define it
[name="pipeline" type="MyModule"]
```

Thanks for tour time!

**Comment by drbenmorgan** [comment]
I've just tested the basic MyModule example and can (modulo some other issues that I'll fixe separately), the problem is due to a grammar change in the pipeline script that wasn't updated. It should be sufficient to change the `flreconstruct.plugins` section header from:

```
[name="flreconstruct.plugins" type=""]
```

to

```
[name="flreconstruct.plugins" type="flreconstruct::section"]
```

Let me know if that works.

**Comment by noeroy** [comment]
That works for me, thanks!

**Comment by cherylepatrick** [comment]
Works for me also, thanks @drbenmorgan

**Comment by drbenmorgan** [comment]
I'll mark this as closed as the fix seems to have worked and PR #17 addresses the documentation issue.
